File: java/src/main/java/io/hatake/core/ReportDefinition.java
package io.hatake.core;

import java.util.List;

/**
 * 帳票（{@code type: report}）の「紙の構造」。明細の列はページの {@code table} から
 * 取るので、一覧と帳票で列がずれない。
 *
 * @param paperSize 用紙サイズ（{@code A4} など。既定 {@code A4}）
 * @param orientation 用紙の向き（{@code portrait} / {@code landscape}）
 * @param rowsPerPage 1枚に載る行数。<b>グループ見出し・小計も1行として数える</b>
 *     （これでページ割りが言語をまたいで一致する）
 * @param groups コントロールブレイク（外側から順。DSL キーは {@code groupBy}）
 * @param totals 小計・総計に出す数字（宣言順）
 * @param limit 1回の出力で読む行数（帳票は「印刷するもの」なのでページングしない）
 * @param sortField 印字順（DSL キーは {@code sort}）。グループはコントロールブレイクなので
 *     並び順が出力に効く。列見出しを押せない帳票では、ここが唯一の指定場所
 * @param sortAscending 昇順か
 */
public record ReportDefinition(
        String paperSize,
        String orientation,
        int rowsPerPage,
        List<ReportGroup> groups,
        List<ReportTotal> totals,
        int limit,
        String sortField,
        boolean sortAscending) {

    public static final String PORTRAIT = "portrait";
    public static final String LANDSCAPE = "landscape";

    /** 既定の帳票（A4 縦・40行・グループ無し）。 */
    public static final ReportDefinition DEFAULT = new ReportDefinition(
            "A4", PORTRAIT, 40, List.of(), List.of(), 1000, null, true);

    public boolean isLandscape() {
        return LANDSCAPE.equals(orientation);
    }
}

File: flutter/packages/hatake_core/lib/src/definition/report_definition.dart
import 'package:equatable/equatable.dart';

import 'aggregate_ops.dart';
import 'paper_definition.dart';

/// A control break: rows whose [field] value changes start a new group.
///
/// Grouping is a **control break over consecutive rows**, exactly like a printed
/// form: the rows must already arrive in the right order (that is the
/// repository's job), so the same value appearing twice apart makes two groups.
class ReportGroup extends Equatable {
  /// Field whose change breaks the group.
  final String field;

  /// Heading label shown next to the group's value.
  final String label;

  /// Start a new sheet whenever this group changes.
  final bool pageBreak;

  const ReportGroup({
    required this.field,
    required this.label,
    this.pageBreak = false,
  });

  @override
  List<Object?> get props => [field, label, pageBreak];
}

/// One figure on the subtotal / grand-total lines.
///
/// Reuses the dashboard's aggregate vocabulary ([AggregateOps]), so `sum` means
/// the same thing on a report as on a metric card. Two totals may share a
/// [field] (e.g. `sum` and `count` of the same column).
class ReportTotal extends Equatable {
  final String field;

  /// Aggregate operation name ([AggregateOps] or a plugin's).
  final String aggregate;

  const ReportTotal({
    required this.field,
    this.aggregate = AggregateOps.sum,
  });

  @override
  List<Object?> get props => [field, aggregate];
}

/// The printing side of a `report` page: paper, how many lines fit on a sheet,
/// and the group / total structure. The detail columns come from the page's
/// `table`, so a report and a list of the same data stay in sync.
class ReportDefinition extends Equatable {
  final PaperDefinition paper;

  /// Lines per sheet. **Group headings and total lines count as lines**, which
  /// is what makes page breaks deterministic across languages.
  final int rowsPerPage;

  /// Control breaks, outermost first.
  final List<ReportGroup> groups;

  /// Figures on the subtotal / grand-total lines, in declaration order.
  final List<ReportTotal> totals;

  /// Rows to fetch for one run. A report is printed, not scrolled, so it reads
  /// one bounded chunk instead of paging.
  final int limit;

  /// Print order, passed to the repository. A report has no clickable headers,
  /// so this is the only place its order can be stated — and [groups] need the
  /// rows to arrive in it.
  final String? sortField;

  final bool sortAscending;

  const ReportDefinition({
    this.paper = const PaperDefinition(),
    this.rowsPerPage = 40,
    this.groups = const [],
    this.totals = const [],
    this.limit = 1000,
    this.sortField,
    this.sortAscending = true,
  });

  @override
  List<Object?> get props =>
      [paper, rowsPerPage, groups, totals, limit, sortField, sortAscending];
}
